lib/backends/webgl/ops/conv-pack.ts

```diff
@@ -0,0 +1,117 @@
+    this.unpackedImpl.initialize(attributes);
+  }
+  protected artifacts: Artifact[];
+  protected programInfo: ProgramInfo[];
+
+  run(inferenceHandler: WebGLInferenceHandler, inputs: Tensor[]): Tensor[] {
+    const programManager = inferenceHandler.session.programManager;
+    const xshape = inputs[0].dims.slice();
+    const fallback =
+        (xshape.length !== 4 || xshape[0] !== 1 || this.group !== 1 ||
+         (this.kernelShape[0] === 1 && this.kernelShape[1] === 1));
+
+    if (fallback || !inferenceHandler.session.pack) {
+      return this.unpackedImpl.run(inferenceHandler, inputs);
+    }
+
+    const kshape = inputs[1].dims.slice();
+    // if kernelShape is not specified in the attributes of this op, infer it from the weight tensor dims
+    if (this.kernelShape.length === 0) {
+      for (let i = 2; i < kshape.length; ++i) {
+        this.kernelShape.push(kshape[i]);
+      }
+    }
+    PoolConvUtil.adjustPadsBasedOnAutoPad(
+        inputs[0].dims, this.strides, this.dilations, this.kernelShape, this.pads, this.autoPad);
+    Logger.verbose(
+        'Conv',
+        `autpPad:${this.autoPad}, dilations:${this.dilations}, group:${this.group}, kernelShape:${
+            this.kernelShape}, pads:${this.pads}, strides:${this.strides}`);
+
+    const outputShape = WebGLConv.calcOutputShape(xshape, kshape, this.dilations, this.pads, this.strides);
+    const im2col = new WebGLIm2ColPacked(outputShape, kshape, this.dilations, this.pads, this.strides);
+    const matmul = new WebGLMatMulPacked();
+    const reshape = new WebGLReshapePacked();
+    // shape for kernel reshape
+    const shape = new Tensor([2], 'int32');
+    shape.data[0] = kshape[0];
+    shape.data[1] = kshape[1] * kshape[2] * kshape[3];
+
+    if (!this.artifacts) {
+      this.artifacts = [];
+      this.programInfo = [];
+      this.programInfo[0] = im2col.createProgramInfo(inferenceHandler, [inputs[0], inputs[1]]);
+      this.artifacts[0] = programManager.build(this.programInfo[0]);
+
+      this.programInfo[1] = reshape.createProgramInfo(inferenceHandler, [inputs[1], shape]);
+      this.artifacts[1] = programManager.build(this.programInfo[1]);
+    }
+
+    // run im2col
+    const runDataIm2col = im2col.createRunData(inferenceHandler, this.programInfo[0], [inputs[0], inputs[1]]);
+    inferenceHandler.checkAndUpdateTextureForm(this.artifacts[0], runDataIm2col);
+    programManager.run(this.artifacts[0], runDataIm2col);
+    const im2colOutput = runDataIm2col.outputTextureData.tensor;
+
+    // reshape kernel
+    const runDataKernelReshape = reshape.createRunData(inferenceHandler, this.programInfo[1], [inputs[1], shape]);
+    inferenceHandler.checkAndUpdateTextureForm(this.artifacts[1], runDataKernelReshape);
+    programManager.run(this.artifacts[1], runDataKernelReshape);
+    const kernelReshaped = runDataKernelReshape.outputTextureData.tensor;
+
+    // run matmul
+    const hasBias = (inputs.length === 3);
+    if (this.artifacts.length === 2) {
```

Comment:
Maybe we should add an assertion instead of an if to ensure at this point, the previous artifacts are created?

Reply:
done.